File: src/security/process-context.ts
/**
 * ProcessContext — Tier E security chokepoint for child-process spawners.
 *
 * Every module that spawns a child process (spawn, exec, execFile, fork,
 * and their *Sync variants) accepts an optional `ProcessContext`. The
 * context defines:
 *
 *   - `allowList` — command patterns the caller is permitted to spawn.
 *     Matching is performed against the executable name/path (the first
 *     argument to spawn/exec — not the full command line). If a command
 *     does not match any pattern, the chokepoint rejects it before the
 *     process is started.
 *   - `audit` — observation sink that receives one record per spawn
 *     attempt (allowed or denied), including the argv vector. Use to wire
 *     central security telemetry.
 *
 * `ProcessContext` is OPTIONAL on every spawn entry point. When absent,
 * callers fall back to the historical permissive behavior so existing call
 * sites continue to work. A `defaultProcessContext()` helper is provided
 * for tests and call sites that want explicit opt-in to the chokepoint
 * with no restrictions (audit-only).
 *
 * Sibling of `LoaderContext` (v1.49.782) and `EgressContext` (this ship);
 * the three chokepoints share a common shape but specialize their targets
 * (filesystem path / URL / command+argv). See `docs/security-chokepoints.md`
 * for the cross-surface catalog and migration policy.
 *
 * Migration policy: new spawn callers MUST accept `ctx?: ProcessContext` and
 * route every spawn/exec/fork invocation through
 * `ensureProcessAllowed(ctx, command, op, argv?)`. Enforced by
 * `tools/security/audit-process-context.mjs` (run by pre-tag-gate) which
 * greps for `node:child_process` imports in `src/` and reports unwired
 * call sites.
 *
 * @module security/process-context
 */

// ============================================================================
// Pattern matching
// ============================================================================

/**
 * A single allow-list entry for child-process commands.
 *
 *   - `string` — matches if `command === pattern` OR `command.startsWith(pattern)`
 *     when `pattern` ends with `/`. Use exact-match for executables on PATH
 *     (e.g. `'git'`) and prefix-match for absolute paths to interpreter dirs
 *     (e.g. `'/usr/local/bin/'`).
 *   - `RegExp` — matches if `pattern.test(command)`.
 *   - `(command: string) => boolean` — custom predicate. Returns truthy → allow.
 *
 * Note: the pattern is checked against the EXECUTABLE (first arg to spawn/exec),
 * NOT the full command line. Argument vetting is the caller's responsibility;
 * the chokepoint records argv in the audit record so downstream telemetry can
 * see it.
 */
export type CommandPattern = string | RegExp | ((command: string) => boolean);

/** Return true when `command` matches any pattern in `allowList`. */
export function matchesCommandAllowList(
  allowList: readonly CommandPattern[],
  command: string,
): boolean {
  for (const pat of allowList) {
    if (typeof pat === 'string') {
      if (pat.endsWith('/')) {
        if (command === pat.slice(0, -1) || command.startsWith(pat)) return true;
      } else if (command === pat) {
        return true;
      }
    } else if (pat instanceof RegExp) {
      if (pat.test(command)) return true;
    } else if (typeof pat === 'function') {
      if (pat(command)) return true;
    }
  }
  return false;
}

// ============================================================================
// Audit sink
// ============================================================================

/** Operation tag attached to process audit records. Extend as new shapes appear. */
export type ProcessOp =
  | 'spawn'
  | 'spawn-sync'
  | 'exec'
  | 'exec-sync'
  | 'exec-file'
  | 'exec-file-sync'
  | 'fork';

/** One row of audit output emitted by a child-process caller. */
export interface ProcessAuditRecord {
  /** Caller module that emitted the record, e.g. `dependency-auditor/dry-run-gate`. */
  source: string;
  /** Operation being performed. */
  op: ProcessOp;
  /** Executable about to be spawned (first arg to spawn/exec). */
  target: string;
  /** Argument vector passed alongside `target`. Empty array for shell-string exec(). */
  argv: readonly string[];
  /** Whether the allowList admitted the operation. */
  allowed: boolean;
  /** Time the record was emitted (Date.now() at call site). */
  timestamp: number;
  /** Optional human-readable note (e.g. cwd, timeout). */
  note?: string;
}

/** Pluggable observation sink. */
export interface ProcessAuditSink {
  record(entry: ProcessAuditRecord): void;
}

/** No-op sink — default when callers don't want audit output. */
export const NULL_PROCESS_AUDIT_SINK: ProcessAuditSink = {
  record(): void {
    /* intentionally empty */
  },
};

/**
 * Capturing sink — collects records in memory.
 * Intended for tests and short-lived diagnostics, not production telemetry.
 */
export class CapturingProcessAuditSink implements ProcessAuditSink {
  readonly records: ProcessAuditRecord[] = [];
  record(entry: ProcessAuditRecord): void {
    this.records.push(entry);
  }
  clear(): void {
    this.records.length = 0;
  }
}

// ============================================================================
// Context shape
// ============================================================================

/**
 * The chokepoint type threaded through every child-process caller.
 *
 * Both fields are required when a context is provided. Pass
 * `defaultProcessContext()` for a permissive-but-audited context, or omit
 * the parameter entirely for historical permissive behavior.
 */
export interface ProcessContext {
  /** Allow-list of command patterns the caller may spawn. Empty list = deny all. */
  readonly allowList: readonly CommandPattern[];
  /** Sink for audit records. */
  readonly audit: ProcessAuditSink;
}

/**
 * Build a permissive context (matches every command) that still emits audit
 * records. Useful for tests and incremental rollout.
 */
export function defaultProcessContext(
  sink: ProcessAuditSink = NULL_PROCESS_AUDIT_SINK,
): ProcessContext {
  return {
    allowList: [/.*/],
    audit: sink,
  };
}

// ============================================================================
// Enforcement helper
// ============================================================================

/** Error thrown when a spawn operation is rejected by a `ProcessContext.allowList`. */
export class ProcessContextDenied extends Error {
  readonly target: string;
  readonly argv: readonly string[];
  readonly source: string;
  readonly op: ProcessOp;
  constructor(source: string, op: ProcessOp, target: string, argv: readonly string[]) {
    super(`ProcessContext denied ${op} of ${target} (source: ${source})`);
    this.name = 'ProcessContextDenied';
    this.source = source;
    this.op = op;
    this.target = target;
    this.argv = argv;
  }
}

/**
 * Gate a single child-process spawn through a `ProcessContext`.
 *
 *   - If `ctx` is `undefined`, returns silently (legacy permissive mode).
 *   - Otherwise checks `ctx.allowList` against `command` and emits one
 *     audit record including the full argv.
 *   - Throws `ProcessContextDenied` when the operation is not allowed.
 *
 * Call this at every entry point that starts a child process, BEFORE
 * the spawn.
 *
 * @param ctx Optional context; `undefined` → legacy permissive mode.
 * @param source Caller module identifier (e.g. `dependency-auditor/dry-run-gate`).
 * @param op Tag describing the spawn variant.
 * @param command Executable name or path about to be spawned.
 * @param argv Argument vector (empty array for shell-string exec).
 * @param note Optional context for the audit record (e.g. cwd, timeout).
 */
export function ensureProcessAllowed(
  ctx: ProcessContext | undefined,
  source: string,
  op: ProcessOp,
  command: string,
  argv: readonly string[] = [],
  note?: string,
): void {
  if (!ctx) return;
  const allowed = matchesCommandAllowList(ctx.allowList, command);
  ctx.audit.record({
    source,
    op,
    target: command,
    argv,
    allowed,
    timestamp: Date.now(),
    ...(note !== undefined ? { note } : {}),
  });
  if (!allowed) {
    throw new ProcessContextDenied(source, op, command, argv);
  }
}

File: src-tauri/src/security/process_context.rs
//! ProcessContext — Rust analogue of the TypeScript Tier-E child-process
//! chokepoint (`src/security/process-context.ts`, v1.49.806).
//!
//! Every `Command::new` / `CommandBuilder::new` site in `src-tauri/src`
//! calls [`ensure_process_allowed`] BEFORE starting the child process. The
//! gate consults an optional process-wide [`ProcessPolicy`]:
//!
//!   - **No policy installed (default)** — legacy permissive mode; the gate
//!     returns `Ok(())` without auditing. This mirrors the TS chokepoint's
//!     optional-`ctx` semantics: enforcement is opt-in, existing behavior is
//!     unchanged until a policy is installed.
//!   - **Policy installed** — the target executable is matched against the
//!     policy's allow-list, one [`ProcessAuditRecord`] is emitted per attempt
//!     (allowed or denied), and a denied attempt returns
//!     [`ProcessContextDenied`] so the spawn never happens.
//!
//! Denial-propagation discipline (Rust variant of #10427): call sites with a
//! `Result` error channel propagate the denial as an error. Detector-shaped
//! sites (`Option`/`bool` returns such as `detect_tmux`, `has_session`) treat
//! denial as fail-closed unavailability — the binary is not available TO US —
//! and the audit record carries the denial signal. The record, not the return
//! value, is the load-bearing security telemetry at those sites.
//!
//! The matching pure core is [`evaluate`]; [`ensure_process_allowed`] is the
//! thin wrapper over the process-wide `OnceLock` policy. Unit tests exercise
//! `evaluate` with explicit policies; exactly ONE test owns the global
//! lifecycle because the `OnceLock` is process-wide and cargo runs tests in
//! parallel threads.
//!
//! Drift-guard: `src/security/process-context-rust-audit.test.ts` (default
//! vitest project → pre-tag-gate + every CI leg) pins the spawner-file roster
//! and requires one `ensure_process_allowed` call per spawn site.

use std::sync::{Arc, Mutex, OnceLock};
use std::time::SystemTime;

/// Spawn variant being gated. Rust's three child-process entry points.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ProcessOp {
    /// `Command::spawn` (also covers portable-pty `spawn_command`).
    Spawn,
    /// `Command::output`.
    Output,
    /// `Command::status`.
    Status,
}

impl std::fmt::Display for ProcessOp {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = match self {
            ProcessOp::Spawn => "spawn",
            ProcessOp::Output => "output",
            ProcessOp::Status => "status",
        };
        f.write_str(s)
    }
}

/// A single allow-list entry for child-process executables.
///
/// Matching runs against the EXECUTABLE (first argument to `Command::new`),
/// not the full command line — argument vetting stays the caller's job; the
/// audit record carries argv for downstream telemetry.
#[derive(Debug, Clone)]
pub enum CommandPattern {
    /// Matches when `target == value` exactly (e.g. `"tmux"`).
    Exact(String),
    /// Matches when `target` starts with `value` (e.g. `"/usr/local/bin/"`).
    Prefix(String),
    /// Matches every target. Used by [`permissive_process_policy`] for
    /// audit-only rollout.
    Any,
}

impl CommandPattern {
    fn matches(&self, target: &str) -> bool {
        match self {
            CommandPattern::Exact(v) => target == v,
            CommandPattern::Prefix(v) => target.starts_with(v.as_str()),
            CommandPattern::Any => true,
        }
    }
}

/// Return true when `target` matches any pattern in `allow_list`.
pub fn matches_command_allow_list(allow_list: &[CommandPattern], target: &str) -> bool {
    allow_list.iter().any(|p| p.matches(target))
}

/// One row of audit output emitted per gated spawn attempt.
#[derive(Debug, Clone)]
pub struct ProcessAuditRecord {
    /// Caller identifier, `module/function` (e.g. `tmux/list_sessions`).
    pub source: String,
    /// Spawn variant being attempted.
    pub op: ProcessOp,
    /// Executable about to be started.
    pub target: String,
    /// Argument vector passed alongside `target`.
    pub argv: Vec<String>,
    /// Whether the allow-list admitted the attempt.
    pub allowed: bool,
    /// Time the record was emitted.
    pub timestamp: SystemTime,
}

/// Pluggable observation sink. Must be `Send + Sync` — the policy holding it
/// is process-wide and gate calls arrive from tokio workers and std threads.
pub trait ProcessAuditSink: Send + Sync {
    fn record(&self, entry: &ProcessAuditRecord);
}

/// Capturing sink — collects records in memory behind an `Arc`, so clones
/// share one buffer (install a clone, keep the original to read). Tests and
/// short-lived diagnostics only.
#[derive(Clone, Default)]
pub struct CapturingProcessAuditSink {
    records: Arc<Mutex<Vec<ProcessAuditRecord>>>,
}

impl CapturingProcessAuditSink {
    pub fn new() -> Self {
        Self::default()
    }

    /// Snapshot of all records captured so far.
    pub fn records(&self) -> Vec<ProcessAuditRecord> {
        self.records.lock().expect("capturing sink poisoned").clone()
    }
}

impl ProcessAuditSink for CapturingProcessAuditSink {
    fn record(&self, entry: &ProcessAuditRecord) {
        self.records
            .lock()
            .expect("capturing sink poisoned")
            .push(entry.clone());
    }
}

/// The chokepoint policy: allow-list + audit sink.
pub struct ProcessPolicy {
    /// Allow-list of executable patterns. Empty list = deny all.
    pub allow_list: Vec<CommandPattern>,
    /// Sink receiving one record per gated attempt. `None` = no audit output.
    pub audit: Option<Box<dyn ProcessAuditSink>>,
}

impl std::fmt::Debug for ProcessPolicy {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("ProcessPolicy")
            .field("allow_list", &self.allow_list)
            .field("audit", &self.audit.is_some())
            .finish()
    }
}

/// Build a permissive policy (admits every target) that still emits audit
/// records — the incremental-rollout / observability mode.
pub fn permissive_process_policy(sink: Box<dyn ProcessAuditSink>) -> ProcessPolicy {
    ProcessPolicy {
        allow_list: vec![CommandPattern::Any],
        audit: Some(sink),
    }
}

/// Error returned when a spawn attempt is rejected by the policy allow-list.
///
/// `caller` is the TS chokepoint's `source` field — renamed here because
/// thiserror reserves a field named `source` for the error-source chain.
#[derive(Debug, thiserror::Error)]
#[error("ProcessContext denied {op} of {target} (source: {caller})")]
pub struct ProcessContextDenied {
    pub caller: String,
    pub op: ProcessOp,
    pub target: String,
    pub argv: Vec<String>,
}

/// Pure gate core: check `target` against an explicit (optional) policy.
///
///   - `None` policy → `Ok(())`, no audit (legacy permissive mode).
///   - `Some` policy → allow-list check + one audit record per attempt;
///     denied attempts return [`ProcessContextDenied`].
pub fn evaluate(
    policy: Option<&ProcessPolicy>,
    source: &str,
    op: ProcessOp,
    target: &str,
    argv: &[&str],
) -> Result<(), ProcessContextDenied> {
    let Some(policy) = policy else {
        return Ok(());
    };
    let allowed = matches_command_allow_list(&policy.allow_list, target);
    if let Some(sink) = &policy.audit {
        sink.record(&ProcessAuditRecord {
            source: source.to_string(),
            op,
            target: target.to_string(),
            argv: argv.iter().map(|a| a.to_string()).collect(),
            allowed,
            timestamp: SystemTime::now(),
        });
    }
    if allowed {
        Ok(())
    } else {
        Err(ProcessContextDenied {
            caller: source.to_string(),
            op,
            target: target.to_string(),
            argv: argv.iter().map(|a| a.to_string()).collect(),
        })
    }
}

static PROCESS_POLICY: OnceLock<ProcessPolicy> = OnceLock::new();

/// Install the process-wide policy. One-shot (`OnceLock` semantics): returns
/// the rejected policy as `Err` when one is already installed. Nothing in the
/// app installs a policy today — call sites run in legacy permissive mode
/// until a startup hook opts in.
pub fn install_process_policy(policy: ProcessPolicy) -> Result<(), ProcessPolicy> {
    PROCESS_POLICY.set(policy)
}

/// Gate a child-process start through the process-wide policy. Call BEFORE
/// every `Command::new(..)` spawn/output/status (and portable-pty
/// `spawn_command`), with `source` as `module/function`.
pub fn ensure_process_allowed(
    source: &str,
    op: ProcessOp,
    target: &str,
    argv: &[&str],
) -> Result<(), ProcessContextDenied> {
    evaluate(PROCESS_POLICY.get(), source, op, target, argv)
}
